exp/query/lsm_fetcher.go

```diff
@@ -0,0 +1,131 @@
+	"github.com/weaviate/weaviate/entities/cyclemanager"
+)
+
+const (
+	defaultLSMRoot = "lsm"
+
+	// NOTE(kavi): using fixed nodeName that offloaded tenant under that prefix on object storage.
+	// TODO(kavi): Make it dynamic.
+	nodeName = "weaviate-0"
+)
+
+// LSMFetcher fetches the tenant's LSM data using given `upstream` without or without local cache.
+type LSMFetcher struct {
+	cache    LSMCache
+	upstream LSMDownloader
+	log      logrus.FieldLogger
+
+	basePath string
+}
+
+// LSMDownloader is usually some component that used to download tenant's data.
+type LSMDownloader interface {
+	DownloadToPath(ctx context.Context, collection, tenant, nodeName, path string) error
+}
+
+// LSMCache provides caching of tenant's LSM data.
+type LSMCache interface {
+	Tenant(collection, tenantID string) (*TenantCache, error)
+	AddTenant(collection, tenantID string, version uint64) error
+	BasePath() string
+}
+
+// NewLSMFetcher creates LSMFetcher without cache.
+// Every call is forwarded to given `upstream` to fetch the LSM data for the tenant.
+func NewLSMFetcher(basePath string, upstream LSMDownloader, log logrus.FieldLogger) *LSMFetcher {
+	return NewLSMFetcherWithCache(basePath, upstream, nil, log)
+}
+
+// NewLSMFetcherWithCache creates LSMFetcher with cache.
+func NewLSMFetcherWithCache(basePath string, upstream LSMDownloader, cache LSMCache, log logrus.FieldLogger) *LSMFetcher {
+	return &LSMFetcher{
+		basePath: basePath,
+		cache:    cache,
+		log:      log,
+		upstream: upstream,
+	}
+}
+
+// Fetch try to fetch LSM data for the given (collection,tenant).
+// It works in following modes
+// 1. If cache is nil, fetch from upstream.
+// 2. If cache is enabled and not found in cache, fetch from upstream.
+// 3. If cache is enabled and found in the cache but local version is outdated, fetch from the upstream
+// 4. if cache is enabled, found in the cache and local version is up to date, return from the cache.
+func (l *LSMFetcher) Fetch(ctx context.Context, collection, tenant string, version uint64) (*lsmkv.Store, string, error) {
+	var (
+		basePath          = l.basePath
+		tenantPath string = ""
+		download   bool   = true
+	)
+
+	if l.cache != nil {
+		basePath = l.cache.BasePath()
+		c, err := l.cache.Tenant(collection, tenant)
+		if err == nil && version != 0 && c.Version >= version {
+			tenantPath = c.AbsolutePath()
+			download = false
+		}
+		if err != nil && !errors.Is(err, ErrTenantNotFound) {
+			return nil, "", err
+		}
+	}
+
+	if download {
+		tenantPath = path.Join(
+			basePath,
+			strings.ToLower(collection),
+			strings.ToLower(tenant),
+		)
+		// src - s3://<collection>/<tenant>/<node>/
+		// dst (local) - <data-path/<collection>/<tenant>
+		l.log.WithFields(logrus.Fields{
+			"collection": collection,
+			"tenant":     tenant,
+			"nodeName":   nodeName,
+			"dst":        tenantPath,
+		}).Debug("starting download to path")
+		if err := l.upstream.DownloadToPath(ctx, collection, tenant, nodeName, tenantPath); err != nil {
```

Comment:
If we're downloading a new version of this tenant, don't we need to remove the old one? Otherwise we would have all the files merged for each version each time we download?

Reply:
Good point. I would do following instead of separate step to remove the old tenant.

Say we have some version under `/cache/collection/tenant-1/` and we found more latest version. All downloads should happen in some tmp location (`/tmp/collection/tenant-1/`) first and atomically replace old with new (`mv /tmp/collection/tenant-1/` `cache/collection/tenant-1/`) using Go's `os.Rename` (which is atomic as long as both source and destination in same filesystem)

That's usually recommended way to do such things.

Rationale being, I don't want to maintain multiple versions of `/collection/tenant-1/` in local cache. That way there will always be one copy of latest tenant data at any point in time. Easy to mental model IMHO.

wdyt?

Reply:
I actually ended up using `/<collection>/<tenant>/<version>` and deleting the old version when updated **after** updating the cache map (so that all new tenants have access to new tenant data). Basically similar to what you proposed.

I'm not 100% sure about what is the right time to delete the old version. Even though every lookup after adding tenant will return new path, still there can be client using old path and deleting can result in client failures. Added TODO about this known constrains and will come back to it if that's causing huge problems.

Reply:
Done